### ISSUE TITLE
print(x.grad): <Tensor <LB (3, 3) op:MovementOps.RESHAPE> with grad None> <Tensor <LB (1, 3) op:MovementOps.RESHAPE> with grad None>

### ISSUE THREAD

**Issue by dh-bccw** [geohot/tinygrad#441]
Environment:
windows10

python:
from tinygrad.tensor import Tensor

x = Tensor.eye(3, requires_grad=True)
y = Tensor([[2.0,0,-2.0]], requires_grad=True)
z = y.matmul(x).sum()
z.backward()

print(x.grad)  # dz/dx
print(y.grad)  # dz/dy

Out:
<Tensor <LB (3, 3) op:MovementOps.RESHAPE> with grad None>
<Tensor <LB (1, 3) op:MovementOps.RESHAPE> with grad None>

!!!Why is that? How to solve this problem? thank you!!!

**Comment by Ziv-spec** [comment]
If you are asking why it does not print out the  gradient it is because you need to ask for it explicitly. 

`x.grad.data` and `y.grad.data` respectively

tinygrad does the computation only when you ask for the data in a lazy manner so for your example:  

```py
>>> from tinygrad.tensor import Tensor
>>> x = Tensor.eye(3, requires_grad=True)
>>> y = Tensor([[2.0,0,-2.0]], requires_grad=True)
>>> z = y.matmul(x).sum()
>>> z.backward()
>>> y.grad.data
array([[1., 1., 1.]], dtype=float32)
>>> x.grad.data
array([[ 2.,  2.,  2.],
       [ 0.,  0.,  0.],
       [-2., -2., -2.]], dtype=float32)
```

**Comment by garrett-partenza-us** [comment]
If you do not want lazy computation you can pass LAZY=0 before executing your script from CLI. For example,

`LAZY=0 python3 myscript.py`

**Comment by dh-bccw** [comment]
> If you are asking why it does not print out the gradient it is because you need to ask for it explicitly.
> 
> `x.grad.data` and `y.grad.data` respectively
> 
> tinygrad does the computation only when you ask for the data in a lazy manner so for your example:
> 
> ```python
> >>> from tinygrad.tensor import Tensor
> >>> x = Tensor.eye(3, requires_grad=True)
> >>> y = Tensor([[2.0,0,-2.0]], requires_grad=True)
> >>> z = y.matmul(x).sum()
> >>> z.backward()
> >>> y.grad.data
> array([[1., 1., 1.]], dtype=float32)
> >>> x.grad.data
> array([[ 2.,  2.,  2.],
>        [ 0.,  0.,  0.],
>        [-2., -2., -2.]], dtype=float32)
> ```



> Environment: windows10
> 
> python: from tinygrad.tensor import Tensor
> 
> x = Tensor.eye(3, requires_grad=True) y = Tensor([[2.0,0,-2.0]], requires_grad=True) z = y.matmul(x).sum() z.backward()
> 
> print(x.grad) # dz/dx print(y.grad) # dz/dy
> 
> Out: <Tensor <LB (3, 3) op:MovementOps.RESHAPE> with grad None> <Tensor <LB (1, 3) op:MovementOps.RESHAPE> with grad None>
> 
> !!!Why is that? How to solve this problem? thank you!!!



> If you are asking why it does not print out the gradient it is because you need to ask for it explicitly.
> 
> `x.grad.data` and `y.grad.data` respectively
> 
> tinygrad does the computation only when you ask for the data in a lazy manner so for your example:
> 
> ```python
> >>> from tinygrad.tensor import Tensor
> >>> x = Tensor.eye(3, requires_grad=True)
> >>> y = Tensor([[2.0,0,-2.0]], requires_grad=True)
> >>> z = y.matmul(x).sum()
> >>> z.backward()
> >>> y.grad.data
> array([[1., 1., 1.]], dtype=float32)
> >>> x.grad.data
> array([[ 2.,  2.,  2.],
>        [ 0.,  0.,  0.],
>        [-2., -2., -2.]], dtype=float32)
> ```

Thank you!

**Comment by dh-bccw** [comment]
> If you do not want lazy computation you can pass LAZY=0 before executing your script from CLI. For example,
> 
> `LAZY=0 python3 myscript.py`

Thank you!
